File: templates/zig/src/anr/rlp.zig
//! Minimal RLP encoder/decoder for ANR wire format.

const std = @import("std");

// ── encode ────────────────────────────────────────────────────────────────────

/// Encode a list of byte slices as an RLP list. Caller frees result.
pub fn encodeList(alloc: std.mem.Allocator, items: []const []const u8) ![]u8 {
    var payload = std.ArrayList(u8).init(alloc);
    defer payload.deinit();

    for (items) |item| {
        const encoded = try encodeBytes(alloc, item);
        defer alloc.free(encoded);
        try payload.appendSlice(encoded);
    }

    var out = std.ArrayList(u8).init(alloc);
    try appendLengthPrefix(&out, payload.items.len, 0xC0);
    try out.appendSlice(payload.items);
    return out.toOwnedSlice();
}

/// Encode a byte slice as an RLP string item. Caller frees result.
pub fn encodeBytes(alloc: std.mem.Allocator, data: []const u8) ![]u8 {
    var out = std.ArrayList(u8).init(alloc);
    if (data.len == 1 and data[0] < 0x80) {
        try out.append(data[0]);
    } else {
        try appendLengthPrefix(&out, data.len, 0x80);
        try out.appendSlice(data);
    }
    return out.toOwnedSlice();
}

fn appendLengthPrefix(out: *std.ArrayList(u8), length: usize, offset: u8) !void {
    if (length < 56) {
        try out.append(offset + @as(u8, @intCast(length)));
    } else {
        var be_buf: [8]u8 = undefined;
        var ll: u8 = 0;
        var n = length;
        while (n > 0) : (n >>= 8) { be_buf[7 - ll] = @intCast(n & 0xFF); ll += 1; }
        try out.append(offset + 55 + ll);
        try out.appendSlice(be_buf[8 - ll ..]);
    }
}

// ── decode ────────────────────────────────────────────────────────────────────

/// Decode RLP-encoded data as a list of byte slices.
/// Each slice is freshly allocated. Caller frees each slice and the ArrayList.
pub fn decodeList(alloc: std.mem.Allocator, data: []const u8) !std.ArrayList([]u8) {
    var list = std.ArrayList([]u8).init(alloc);
    errdefer {
        for (list.items) |item| alloc.free(item);
        list.deinit();
    }

    const prefix = data[0];
    var payload: []const u8 = undefined;

    if (prefix <= 0xF7) {
        const len = prefix - 0xC0;
        payload = data[1 .. 1 + len];
    } else {
        const ll  = prefix - 0xF7;
        var len: usize = 0;
        for (data[1 .. 1 + ll]) |b| len = (len << 8) | b;
        payload = data[1 + ll .. 1 + ll + len];
    }

    var pos: usize = 0;
    while (pos < payload.len) {
        const p = payload[pos];
        if (p < 0x80) {
            const item = try alloc.alloc(u8, 1);
            item[0] = p;
            try list.append(item);
            pos += 1;
        } else if (p <= 0xB7) {
            const len   = p - 0x80;
            const item  = try alloc.dupe(u8, payload[pos + 1 .. pos + 1 + len]);
            try list.append(item);
            pos += 1 + len;
        } else if (p <= 0xBF) {
            const ll    = p - 0xB7;
            var len: usize = 0;
            for (payload[pos + 1 .. pos + 1 + ll]) |b| len = (len << 8) | b;
            const item  = try alloc.dupe(u8, payload[pos + 1 + ll .. pos + 1 + ll + len]);
            try list.append(item);
            pos += 1 + ll + len;
        } else {
            return error.NestedListNotSupported;
        }
    }

    return list;
}

File: templates/rust/src/anr/rlp.rs
/// Minimal RLP encoder/decoder — sufficient for ANR wire format.

#[derive(Debug, Clone, PartialEq)]
pub enum RlpItem {
    Bytes(Vec<u8>),
    List(Vec<RlpItem>),
}

// ── encode ────────────────────────────────────────────────────────────────────

pub fn rlp_encode(item: &RlpItem) -> Vec<u8> {
    match item {
        RlpItem::Bytes(b) => encode_bytes(b),
        RlpItem::List(l)  => {
            let payload: Vec<u8> = l.iter().flat_map(rlp_encode).collect();
            let mut out = encode_length(payload.len(), 0xC0);
            out.extend(payload);
            out
        }
    }
}

fn encode_bytes(b: &[u8]) -> Vec<u8> {
    if b.len() == 1 && b[0] < 0x80 {
        return b.to_vec();
    }
    let mut out = encode_length(b.len(), 0x80);
    out.extend_from_slice(b);
    out
}

fn encode_length(len: usize, offset: u8) -> Vec<u8> {
    if len < 56 {
        vec![offset + len as u8]
    } else {
        let be: Vec<u8> = {
            let mut n = len;
            let mut bytes = Vec::new();
            while n > 0 { bytes.push(n as u8); n >>= 8; }
            bytes.reverse();
            bytes
        };
        let mut out = vec![offset + 55 + be.len() as u8];
        out.extend(be);
        out
    }
}

// ── decode ────────────────────────────────────────────────────────────────────

#[derive(Debug)]
pub struct RlpError(pub String);
impl std::fmt::Display for RlpError { fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result { write!(f, "{}", self.0) } }
impl std::error::Error for RlpError {}

pub fn rlp_decode(data: &[u8]) -> Result<RlpItem, RlpError> {
    let (item, _) = decode_at(data, 0)?;
    Ok(item)
}

fn decode_at(data: &[u8], offset: usize) -> Result<(RlpItem, usize), RlpError> {
    let prefix = *data.get(offset).ok_or_else(|| RlpError("unexpected end".into()))?;

    if prefix < 0x80 {
        return Ok((RlpItem::Bytes(vec![prefix]), offset + 1));
    }
    if prefix <= 0xB7 {
        let len   = (prefix - 0x80) as usize;
        let start = offset + 1;
        return Ok((RlpItem::Bytes(data[start..start + len].to_vec()), start + len));
    }
    if prefix <= 0xBF {
        let ll    = (prefix - 0xB7) as usize;
        let len   = usize_from_be(&data[offset + 1..offset + 1 + ll]);
        let start = offset + 1 + ll;
        return Ok((RlpItem::Bytes(data[start..start + len].to_vec()), start + len));
    }
    if prefix <= 0xF7 {
        let len   = (prefix - 0xC0) as usize;
        let start = offset + 1;
        let items = decode_list(&data[start..start + len])?;
        return Ok((RlpItem::List(items), start + len));
    }
    let ll    = (prefix - 0xF7) as usize;
    let len   = usize_from_be(&data[offset + 1..offset + 1 + ll]);
    let start = offset + 1 + ll;
    let items = decode_list(&data[start..start + len])?;
    Ok((RlpItem::List(items), start + len))
}

fn decode_list(data: &[u8]) -> Result<Vec<RlpItem>, RlpError> {
    let mut items = Vec::new();
    let mut pos = 0;
    while pos < data.len() {
        let (item, next) = decode_at(data, pos)?;
        items.push(item);
        pos = next;
    }
    Ok(items)
}

fn usize_from_be(b: &[u8]) -> usize {
    b.iter().fold(0usize, |acc, &x| (acc << 8) | x as usize)
}
